pytorch_lightning/trainer/trainer.py

```diff
@@ -431,10 +431,12 @@ def fit(
 
         """
         Trainer._log_api_event("fit")
+
         # we reuse fit for other functions. When already set, it shouldn't be modified.
-        if not self.state.running:
-            self.state = TrainerState.FITTING
-        if self._running_stage is None or self.tuning:
+        if not self.state.fn:
+            self.state.fn = TrainerFn.FITTING
+            self.state.status = TrainerStatus.RUNNING
+        if self.state.stage is None or self.tuning:
             self.training = True
 
```

Comment:
should we have a `_fit_impl` function to avoid this re-entry check? this way the top-level entry points set the state first

Reply:
Yes, this would look cleaner. I'll update it